src/devices/bus/isa/sblaster_lle.cpp

```diff
@@ -0,0 +1,381 @@
+		{
+			case 1: m_isa->irq7_w(CLEAR_LINE); break;
+			case 2: m_isa->irq5_w(CLEAR_LINE); break;
+			case 4: m_isa->irq3_w(CLEAR_LINE); break;
+			case 8: m_isa->irq2_w(CLEAR_LINE); break;
+		}
+		m_irq_raised = false;
+	}
+}
+
+// YM3812
+uint8_t isa8_sblaster_2_0_lle_device::ym3812_16_r(offs_t offset)
+{
+	uint8_t retVal = 0xff;
+	switch(offset)
+	{
+		case 0 : retVal = m_ym3812->status_r(); break;
+	}
+	return retVal;
+}
+
+void isa8_sblaster_2_0_lle_device::ym3812_16_w(offs_t offset, uint8_t data)
+{
+	switch(offset)
+	{
+		case 0 : m_ym3812->address_w(data); break;
+		case 1 : m_ym3812->data_w(data); break;
+	}
+}
+
+// 80C51
+/*
+    DSP port 2
+    Bit 0: w OUTPUT_EN: 1 = Mute, 0 = Not M
+    Bit 5: r MIC_COMP:  1 = Mic <- DAC. 0 = Mic -> DAC.
+    Bit 6: r DAV_PC:    1 = Data waiting in buffer for host PC to read it.
+    Bit 7: r DAV_DSP:   1 = Data waiting in buffer for DSP to read it.
+
+*/
+uint8_t isa8_sblaster_2_0_lle_device::dsp_port2_r()
+{
+	uint8_t data = 0;
+
+	data |= (m_dav_dsp ? (1 << 7) : 0);
+	data |= (m_dav_pc  ? (1 << 6) : 0);
+
+	return data;
+}
+void isa8_sblaster_2_0_lle_device::dsp_port2_w(uint8_t data)
+{
+	// Bit 0 is the only output, which mutes audio output when pulled high.
+	m_speaker->set_output_gain(ALL_OUTPUTS, !BIT(data, 0));
+}
+
+/*
+    DSP port 3
+
+    Bit 0: r MIDI RX    MIDI receive data in.
+    Bit 1: w MIDI TX    MIDI transmit data out.
+    Bit 2: w IREQUEST   Strobe low to send IRQ to host PC.
+    Bit 3: w DSP_BUSY   Set high to tell PC we are busy.
+    Bit 4: w DMA_EN#    Set low to enable DMA requests.
+    Bit 5: w DREQUEST   Strobe high to set DRQ.
+    Bit 6: w DSP_WRITE# Strobe low to write a byte to the PC.
+    Bit 7: w DSP_READ#  Strobe low to read a byte from the PC.
+*/
+uint8_t isa8_sblaster_2_0_lle_device::dsp_port3_r()
+{
+	uint8_t data = 0;
+
+	if(m_dma_en) data |= (1 << 4);
+	if(m_drequest) data |= (1 << 5);
+
+	return 0 | data; // latched bits
+}
+void isa8_sblaster_2_0_lle_device::dsp_port3_w(uint8_t data)
+{
+	m_dma_en = BIT(data, 4); // active low
+
+	if(!BIT(data, 2) && m_irequest) // Strobed high -> low
+	{
+		raise_irq();
+	}
+
+	if(BIT(data, 5))
+	{
+		if(m_dma_en == false && m_drequest == false)
+		{
+			raise_dma();
+		}
+	}
+
+	if(!BIT(data, 6))
+	{
+		m_dav_pc = true;
+	}
+
+	if(!BIT(data, 7))
+	{
+		m_dav_dsp = true;
+	}
+
+	m_irequest = BIT(data, 2);
+	m_drequest = BIT(data, 5);
+}
+
+void isa8_sblaster_2_0_lle_device::dsp_reset_w(offs_t offset, uint8_t data)
+{
+	if(BIT(data, 0))
+	{
+		// Reset DSP.
+		m_dsp->reset();
+
+		// Clear all flags.
+		m_irq_in_flag = false;
+		m_dav_dsp = false;
+		m_dav_pc = false;
+		m_irequest = true;  // active low
+		m_drequest = false; // active high
+
+		lower_irq();
+		lower_dma();
+	}
+}
+
+uint8_t isa8_sblaster_2_0_lle_device::dsp_rbuf_status_r(offs_t offset)
+{
+	// 0x22E
+	lower_irq();                        // Reading this bit clears the IRQ.
+	return (m_dav_pc ? 0x80 : 0x00);    // Return the data waiting flag in bit 7.
```

Comment:
You need to guard against debugger accesses by checking `machine().side_effects_disabled()` before doing things with side effects in read handlers.